### PR TITLE
Split minion job for limiting screenshots and run it less often

### PR DESCRIPTION
**Pull request by Martchus**
Fixes: [poo#55922](https://progress.opensuse.org/issues/55922)

### CODE REVIEW

**Comment by Martchus** [comment]
I'll create a new PR which only has the first 2 commits. These commits are ready as-is and would already improve the situation a little bit.

The WIP commit for splitting the Minion job still needs some testing and tweaking. However, @andrii-suse came up with an alternative idea which he'll point out in https://progress.opensuse.org/issues/55922. So it would make sense to discuss it first.

**Comment by Martchus** [comment]
I've tested this locally with OSD data and it works. The "prevent enqueuing new limit_screenshot if there are still inactive/delayed ones" case works as well although I'm not sure whether there's a better way to handle this within the Minion framework.

I also added a unit/integration test which covers the job splitting and especially checks the ID ranges for correctness.

File: lib/OpenQA/Task/Job/Limit.pm [hunk 1]
```diff
@@ -19,22 +19,34 @@ use Mojo::Base 'Mojolicious::Plugin';
...
 use File::Basename qw(basename dirname);
 use OpenQA::Utils qw(:DEFAULT imagesdir);
+use Scalar::Util 'looks_like_number';
+use List::Util 'min';
+
+# define default parameter for batch processing
+use constant SCREENSHOTS_PER_BATCH            => 200000;       # number of screenshots processed in one DB query
+use constant AVERAGE_EXECUTION_TIME_PER_BATCH => 8;            # estimation for SCREENSHOTS_PER_BATCH
+use constant DESIRED_MINION_JOB_RUNTIME       => (60 * 60);    # aim for one hour per Minion job
+use constant BATCHES_PER_MINION_JOB => (DESIRED_MINION_JOB_RUNTIME / AVERAGE_EXECUTION_TIME_PER_BATCH);
```
Comment (kraih) [inline lib/OpenQA/Task/Job/Limit.pm]: This does look a little overcomplicated, but i don't mind it too much.

File: lib/OpenQA/Task/Job/Limit.pm [hunk 2]
```diff
@@ -19,22 +19,34 @@ use Mojo::Base 'Mojolicious::Plugin';
...
 
 sub register {
     my ($self, $app) = @_;
-    $app->minion->add_task(limit_results_and_logs => sub { _limit($app, @_) });
+    my $minion = $app->minion;
+    $minion->add_task(limit_results_and_logs => sub { _limit($app, @_) });
+    $minion->add_task(limit_screenshots      => sub { _limit_screenshots($app, @_) });
```
Comment (kraih) [inline lib/OpenQA/Task/Job/Limit.pm]: `$app` is also available as `$job->app`. There is no need to capture it in the closure.
```suggestion
    $minion->add_task(limit_results_and_logs => \&_limit);
    $minion->add_task(limit_screenshots      => \& _limit_screenshots);
```

File: t/42-screenshots.t [hunk 3]
```diff
@@ -66,9 +69,14 @@ is_deeply(
...
-# limit job results (which involves deleting unused screenshots)
+# limit screenshots
+my %args = (
+    min_screenshot_id     => $screenshots->search(undef, {rows => 1, order_by => {-asc  => 'id'}})->first->id,
+    max_screenshot_id     => $screenshots->search(undef, {rows => 1, order_by => {-desc => 'id'}})->first->id,
+    screenshots_per_batch => OpenQA::Task::Job::Limit::SCREENSHOTS_PER_BATCH,
+);
 combined_like(
-    sub { OpenQA::Task::Job::Limit::_limit($t->app); },
+    sub { OpenQA::Task::Job::Limit::_limit_screenshots($t->app, undef, \%args); },
```
Comment (kraih) [inline t/42-screenshots.t]: There are much better ways to run a job in tests than to call a private method.

Comment (kraih) [reply t/42-screenshots.t]: You might even be able to use `run_gru_job` from our test utils. https://github.com/os-autoinst/openQA/blob/324e1125073b62a38d45cd50679309e599c767bc/t/lib/OpenQA/Test/Utils.pm#L461

File: t/42-screenshots.t [hunk 4]
```diff
@@ -86,6 +95,52 @@ subtest 'screenshots are unique' => sub {
...
+    combined_like(
+        sub { $worker->dequeue(0, {id => $_})->perform for @enqueued_job_ids },
+        qr/removing screenshot/,
+        'screenshots being removed'
+    );
+    $worker->unregister;
+
+    subtest 'enqueued sub jobs have actually removed the screenshots' => sub {
+        is($screenshots->search({filename => 'test-' . $_})->count, 0, "test screenshot $_ removed again")
+          for (000 .. 200);
```
Comment (okurz) [inline t/42-screenshots.t]: where does this magic number come from? Not sure how long the test takes right now but in my experience Test::More calls in a big for loop can take comparably long time. Within the scalability test we also used a different approach. Maybe the same is good here as well

Comment (kraih) [reply t/42-screenshots.t]: Number does seem a bit arbitrary, considering the max_id above is 206.

Comment (Martchus) [reply t/42-screenshots.t]: The 200 corresponds to the screenshot name and is consistent with the loop where those screenshots are inserted. I have nevertheless added a better check instead which doesn't need that number.

File: cpanfile [hunk 5]
```diff
@@ -39,7 +39,7 @@ requires 'Cpanel::JSON::XS';
 requires 'JavaScript::Minifier::XS', '>= 0.11';
 requires 'JSON::Validator';
 requires 'LWP::UserAgent';
-requires 'Minion', '>= 10.0';
+requires 'Minion', '>= 10.04';
```
Comment (kalikiana) [inline cpanfile]: What's the motivation for this? Is it required? Does it bring in features we'll want to make use of?

Comment (kraih) [reply cpanfile]: Yes, it's required for one of the new Minion features used.

**Comment by okurz** [comment]
@Martchus I am worried about the reduced coverage and the incomplete patch coverage. Could it be that we now run code only in subprocesses which aren't covered anymore by coverage analysis? Maybe you need https://github.com/os-autoinst/openQA/blob/master/t/14-grutasks.t#L105

**Comment by Martchus** [comment]
You're right, this is actually performing the jobs within a sub process. So I've added the code you've mentioned. I made it a test utility but I haven't changed the occurrence in the GRU tasks test because in this case the usage might need to change (https://github.com/os-autoinst/openQA/pull/2801). (In my test only a few jobs are executed and some are covering different code paths.)

**Comment by okurz** [comment]
code looks ok but the coverage analysis looks at least flaky. Not sure if all is still covered. Can you please crosscheck that? At least retrigger and see if the same changes come up.

**Comment by Martchus** [comment]
I think the coverage is now correctly computed. And yes, I'm missing one branch but I didn't want to make the test more complicated at this point. (I've tested it manually.)

**Pull request merged**